src/NodaTime/DateInterval.cs

```diff
@@ -167,5 +186,70 @@ public void Deconstruct(out LocalDate start, out LocalDate end)
             start = Start;
             end = End;
         }
+
+        /// <summary>
+        /// Checks whether the given interval has days in common with the current interval.
+        /// </summary>
+        /// <param name="interval">The <see cref="DateInterval"/> to check for intersection with this interval.</param>
+        /// <returns>
+        /// <c>true</c> if <paramref name="interval"/> has days in common with this interval; <c>false</c> otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// The start and end dates of <paramref name="interval" /> are not in the same calendar
+        /// as the start and end date of this interval.
+        /// </exception>
+        public bool Intersects([NotNull]DateInterval interval)
+        {
+            Preconditions.CheckNotNull(interval, nameof(interval));
+
+            Preconditions.CheckArgument(
+                interval.Start.Calendar.Equals(Start.Calendar), nameof(interval),
+                "The start and end dates of the interval to check must be in the same calendar as the start and end dates of this interval.");
+
+            return ContainsExtreme(interval) || interval.ContainsExtreme(this);
+        }
+
+        /// <summary>
+        /// Returns the intersection between the given interval and this interval.
+        /// </summary>
+        /// <param name="interval">
+        /// The specified interval to which return the intersection with.
+        /// </param>
+        /// <returns>
+        /// A <see cref="DateInterval"/> corresponding to the intersection between the given interval and the current
+        /// instance. If there is no intersection, a null reference is returned.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// The start and end dates of <paramref name="interval" /> are not in the same calendar
+        /// as the start and end date of this interval.
+        /// </exception>
+        [CanBeNull]
+        public DateInterval Intersection([NotNull]DateInterval interval)
```

Comment:
It feels like we could get rid of `Intersects` and just make this return null if there's no intersection.

(I hadn't spotted that DateInterval is a class where Interval is a struct. I can't remember offhand why we did that.)

Reply:
Ok, I have a question here about nullability and the use of the attributes.

What exactly are they used for in Noda? I'm aware that there are tests that break when I don't use an attribute where I was supposed to, but is this it?

Do some of the CI tools perform statical analysis such that I get a broken build or warning in cases like the following:

```
var x = obj.SomeMethodThatCanReturnNull();
x.FooBar(); // oops, risk of NRE
```
(Sorry if this is outside the scope of the PR, but I thought it'd make sense to ask here.)

Reply:
Currently, there's very little checking done. I did start writing a Roslyn analyzer, but it ended up being very, very hard to do with any precision. Resharper knows about the attributes and can warn on them though.

We use NotNull for parameters to automate exception documentation.

When C# 8 comes out, hopefully a lot of this can be more obvious from the types, and checked more by the compiler. Until then, it's mostly informational.